app/controllers/products_controller.rb

```diff
@@ -0,0 +1,144 @@
+class ProductsController < ApplicationController
+before_action :find_product, only: [:show, :edit, :update, :retire]
+before_action :find_merchant, except: [:index, :destroy]
```

Comment:
You don't assign `@merchant` on the index method, so the check on line 8 would never happen anyway and `@merchant` will always be not truthy. :) Also, you don't have a destroy action in this controller?